crates/lib/src/qvm/api.rs

```diff
@@ -0,0 +1,141 @@
+//! This module provides functions to for makings calls to the QVM.
+use std::borrow::Cow;
+use std::str::FromStr;
+
+use qcs_api_client_common::ClientConfiguration;
+use quil_rs::{
+    instruction::{ArithmeticOperand, Instruction, MemoryReference, Move},
+    Program,
+};
+
+use super::{Error, QvmResultData};
+use crate::{
+    executable::Parameters,
+    qvm::{Request, Response},
+};
+
+/// Execute a program on the QVM.
+pub async fn run(
+    quil: &str,
+    shots: u16,
+    readouts: &[Cow<'_, str>],
+    params: &Parameters,
+    config: &ClientConfiguration,
+) -> Result<QvmResultData, Error> {
+    #[cfg(feature = "tracing")]
+    tracing::debug!(
+        %shots,
+        ?readouts,
+        ?params,
+        "preparing program to run on QVM"
+    );
+
+    let program = Program::from_str(quil).map_err(Error::Parsing)?;
+
+    run_program(&program, shots, readouts, params, config).await
+}
+
+pub(crate) async fn run_program(
+    program: &Program,
+    shots: u16,
+    readouts: &[Cow<'_, str>],
+    params: &Parameters,
+    config: &ClientConfiguration,
+) -> Result<QvmResultData, Error> {
+    if shots == 0 {
+        return Err(Error::ShotsMustBePositive);
+    }
+
+    // Create a clone of the program so MOVE statements can be prepended to it
+    let mut program = program.clone();
+
+    for (name, values) in params {
+        match program.memory_regions.get(name.as_ref()) {
+            Some(region) => {
+                if region.size.length != values.len() as u64 {
+                    return Err(Error::RegionSizeMismatch {
+                        name: name.clone(),
+                        declared: region.size.length,
+                        parameters: values.len(),
+                    });
+                }
+            }
+            None => {
+                return Err(Error::RegionNotFound { name: name.clone() });
+            }
+        }
+        for (index, value) in values.iter().enumerate() {
+            program.instructions.insert(
+                0,
```

Comment:
I guess it doesn't matter what order the `MOVE` instructions are in, as long as they are all prepended right?

Reply:
Yup, good case to call out, but order doesn't matter for this.